### ISSUE TITLE
[P1][1.8.x][packaging][decision] De quem é o interpretador no pacote nativo? — bloqueia #1403, e decide se no-GIL alcança o canal nativo

### ISSUE THREAD

**Issue by FabioLeitao** [DataBoar/data-boar#1406]
## Decisão bloqueante

Antes de escrever o manifesto nfpm do **#1403**, é preciso decidir **de quem é o interpretador Python** dentro do pacote nativo. A escolha muda tamanho, matriz de build, manutenção e — o ponto decisivo — **se o no-GIL é alcançável por esse canal**.

Não exige código para ser tomada. Exige só ser tomada antes, porque refazer depois joga fora o manifesto e os 5 formatos junto.

## Por que a faixa de versão é obrigatoriamente estreita

Os wheels vendored da camada 1 (numpy, pandas, scipy, scikit-learn) **não são abi3**: são `cp313-cp313-manylinux…`. Um wheel cp313 **não carrega** em 3.14. Só o `boar_fast_filter` é abi3 (`cp38-abi3`, ver #782) e atravessa minors.

Consequência: um pacote com camada 1 vendored fica preso a **exatamente um minor** — `>= 3.X, << 3.(X+1)`, onde X é o minor de build. Não é escolha conservadora, é ABI.

## O argumento que decide: no-GIL

**Nenhuma distro empacota `python3.14t`.** Medido no lab: T14 (LMDE 7 / Debian 13 trixie) = **3.13.5**.

Se o pacote depender do `python3` da distro:

- fica travado no minor que a distro ship
- construir para 3.14 = **não instala em Debian 13**, que é o alvo principal
- **`cp314t` fica inalcançável por este canal, para sempre**

E `cp314t` não é detalhe: é capacidade de tier Enterprise, e mediu **8,1% mais rápido com paridade exata de achados** no benchmark de 2026-07-30 (ver #1398). Entregar pacote nativo que estruturalmente não pode rodar no-GIL é entregar ao cliente Enterprise um canal de segunda classe — o mesmo defeito que #1399 documentou no container, por outra porta.

### (a) Embarcar o interpretador

O pacote carrega o próprio CPython (como a imagem já faz), em `/usr/lib/data-boar/`.

- ✅ imune a `apt upgrade python3` — o bump de minor da distro é irrelevante
- ✅ **`cp314` e `cp314t` disponíveis em qualquer distro**, inclusive as antigas
- ✅ `pin nunca afrouxa` perfeito; camada 1 idêntica byte a byte em toda distro → o **invariante de correção** vira propriedade estrutural, não esperança
- ✅ um build por `(libc × arch)`, não por minor de distro — matriz **menor**, não maior
- ❌ ~150-300 MB por pacote
- ❌ distro nenhuma aceita upstream — fecha a porta de Debian proper / aports / void-packages

### (b) Depender do `python3` da distro

- ✅ pacote pequeno (~MB), idiomático, é o que mantenedor de distro espera
- ✅ único caminho viável para upstream
- ❌ faixa dura `>= 3.X, << 3.(X+1)` e **um build por minor por distro** — matriz que cresce sozinha
- ❌ quebra quando a distro sobe de minor, até você republicar
- ❌ **no-GIL inalcançável**
- ❌ correção passa a depender do numpy que a distro empacotou — tensão direta com o invariante de correção

### (c) Rebuild no postinst (estilo dkms)

- ❌ exige toolchain e/ou rede na máquina do cliente; **mata o air-gap**, que é justamente o caso de uso que justifica pacote nativo. Descartada, registrada para não voltar.

## Recomendação

**(a) para o canal comercial/Enterprise e air-gapped** — é o único que entrega no-GIL e o único em que o invariante de correção é estrutural.
**(b) para o canal comunidade**, se e quando houver ambição de upstream — e aí sim assumindo que ele não carrega no-GIL, **documentado**, não descoberto pelo cliente.

Não é contradição: PyPI e nativo já são canais distintos com propriedades distintas. O erro seria fingir que um pacote serve aos dois propósitos.

## Aceite

- [ ] Decisão registrada em **ADR** (não só em issue): de quem é o interpretador, por canal
- [ ] Se (a): o interpretador embarcado entra no escopo do `EXTRAS_MANIFEST` (#1401) e do manifesto de integridade
- [ ] Se (b): documentar explicitamente que o canal **não** oferece no-GIL, e qual minor cada build alvo
- [ ] `docs/plans/PLAN_NATIVE_PACKAGES.md` atualizado com a decisão antes de qualquer manifesto nfpm
- [ ] **Bloqueia #1403** até resolver

## O que NÃO é problema (verificado)

A **âncora de integridade já cobre upgrade legítimo** desde `1.7.4.post6` — `core/integrity_anchor.py:21`: *"If `release_label` changed (legitimate package upgrade) → re-baseline"*. Vale para `apt`/`dnf` igual vale para `pipx upgrade`, sem nada novo. **Não reimplementar.**

Única borda a tratar: deb/rpm têm **revisão de pacote** separada da versão upstream (`1.8.0-1` → `1.8.0-2`). Rebuild com wheels vendored diferentes mantém `_package_version() == "1.8.0"` → hashes mudam sem mudar o label → **false-tint**. O `pipx` não tem esse conceito, então o post6 não podia prever. Correção pequena em `_release_label()`: incorporar a revisão quando existir.

Relacionadas: #1403 · #1398 (no-GIL medido) · #1401 · #782 (abi3) · #1182 (wheelhouse) · #856 / `PLAN_BUILD_IDENTITY_RELEASE_INTEGRITY.md`

**Comment by FabioLeitao** [comment]
## ⚠️ Correção de uma afirmação minha neste issue

Escrevi *"Nenhuma distro empacota `python3.14t`"*. **Está errado** — eu deduzi do T14 (3.13.5) em vez de medir. Medido agora:

```
fedora:43   python3.14-freethreading.x86_64
            python3.14-freethreading-debug.x86_64
            (+ variantes i686)
```

**Fedora 43 empacota free-threading.** A afirmação correta é: *a maioria das distros ainda não empacota, e o alvo principal (Debian 13 trixie / LMDE 7) tem 3.13.5* — não "nenhuma".

Isso **não muda a recomendação**, e reforça um ponto: depender do `python3` da distro tornaria o no-GIL disponível **só no Fedora 43+**, e indisponível em Debian, Alpine, Void e Arch. Um canal que entrega capacidade Enterprise em uma distro e não nas outras é pior que um canal que entrega em todas.

## Decisão do operador: **3.14 free-threaded**

Fica a opção **(a) — embarcar o interpretador**. Consequências que simplificam o desenho:

- **Não existe `Depends: python3`.** Some a discussão de faixa (`>= 3.12`? `>= 3.13`? teto?) — a distro tem o que tiver.
- Um build por `(libc × arch)`, **não** por minor de distro. Matriz menor.
- `cp314t` em toda distro, inclusive antigas.
- Camada 1 byte a byte idêntica em todo lugar → o **invariante de correção** vira propriedade estrutural.

## Por que o teto era obrigatório na opção (b) — para o registro

Não era conservadorismo. Wheel vendored `cp313-cp313` não carrega em 3.14; o Python procura `…cpython-314-….so` e não acha. E `core/detector.py` protege os imports de ML com `except ImportError` → o scan **cai para regex-only sem avisar**, 13 achados em vez de 26. Falha silenciosa que degrada a correção = **garantia falsa de conformidade**. Por isso `>= 3.13` sem teto não era opção.

Com o interpretador embarcado, esse modo de falha deixa de existir.

**Comment by FabioLeitao** [comment]
## Decisão fechada pelo operador: **um pacote, `cp314t`, interpretador embarcado**

Com dois dados novos, medidos hoje, que mudam o cálculo.

### 1. Embarcar o interpretador é barato — eu tinha errado a estimativa

Medido dentro das imagens publicadas:

```
cp314  (GIL)      interpretador + stdlib    25,6 MB
                  site-packages (wheels)   503,7 MB
                  scipy 76 · scipy.libs 39 · numpy.libs 39 · pandas 38 · sklearn 29 · matplotlib 23

cp314t (no-GIL)   interpretador + stdlib    22,4 MB
                  site-packages (wheels)   516,5 MB
```

Eu havia estimado 150-300 MB para embarcar o CPython. **São ~25 MB.** O peso é a stack científica (~500 MB), que é a camada 1 e vai no pacote de qualquer forma, porque é ela que define a verdade do manifesto.

**O argumento de tamanho contra embarcar o interpretador não existe.**

### 2. A cobertura cp314t já é completa — eu tinha lido errado o wheelhouse

```
cp314  (GIL)   : 145 pacotes
cp314t (no-GIL): 144 pacotes
diferença      : pip  (removido de propósito)
```

Zero pacotes faltando. Meu "46 células GIL contra 10 cp314t" do #1182 estava mal interpretado: o wheelhouse hospeda **apenas o que a PyPI não cobre** naquela célula. numpy, scipy, pandas e sklearn publicam cp314t a montante — as 10 células são os **buracos**, não o total.

### 3. Por que só no-GIL, e não dois stacks

| | dois stacks | **só no-GIL** |
|---|---|---|
| pacotes de núcleo | 2 | **1** |
| ABIs por pacote de extra | 2 | **1** |
| matriz de build | `(libc × arch) × 2` | **`(libc × arch)`** |

E é **doutrinariamente mais correto que gatear o runtime por tier.** O invariante aditivo diz que licença **adiciona capacidade** e que claim de quantidade escala sem remover. Runtime free-threaded para todos; a licença regula o **teto de workers**.

### 4. O clamp de licença foi validado nas duas direções (2026-07-30)

Mesmo config, 8 alvos idênticos, `max_workers: 8`, execuções serializadas:

```
A) enforced + lic enterprise
   license_state=VALID, mode=enforced
   (nenhuma linha de clamp)                                        real  1m20,480s

B) open, sem licença
   license_state=OPEN, mode=open
   LICENSE_AUDIT action=workers_clamped mode=open allowed=false
     state=OPEN detail=requested=8 cap=2 (open-mode clamp #856)     real  1m56,163s
```

O clamp **existe, dispara, é auditável e mede** — e se declara no log em vez de degradar em silêncio, que é a postura correta. Ressalva honesta sobre o número: 44% e não os ~4× que a aritmética ingênua sugeriria, porque a carga é filesystem local com I/O compartilhado. O que valida a decisão é o comportamento, não a razão exata.

Portanto Community roda o mesmo `cp314t` com teto 2; Enterprise roda o mesmo binário com ∞. Sem crippleware, e sem precisar manter uma build GIL só para servir de degrau.

## Consequências para o desenho

- **Não existe `Depends: python3`** — some a discussão de faixa (`>= 3.12`? `>= 3.13`? teto?)
- Um build por `(libc × arch)`, **não** por minor de distro
- Camada 1 byte a byte idêntica em toda distro → **invariante de correção vira propriedade estrutural**

## Duas guardas obrigatórias

1. **O `assert` de build vira gate permanente em todos os artefatos.** `Dockerfile.nogil:69` já falha se o GIL religar ou se sobrar `.so` do SQLAlchemy. Sendo no-GIL a **única** build, qualquer dependência futura que religue o GIL degrada **todo mundo de uma vez**. Esse assert deixa de ser detalhe da imagem e passa a ser gate de release.
2. **Manter a variante GIL construível, mesmo sem publicar.** Se um cliente esbarrar em bug específico de free-threading, é preciso ter rota de fuga. Não publicar ≠ não conseguir construir.

**Comment by FabioLeitao** [comment]
**Decisão do operador (HITL): opção (a) — embarcar o interpretador, com no-GIL (`cp314t`).**

Canal **nativo/Enterprise** carrega o próprio CPython (`/usr/lib/data-boar/`) → `cp314t` alcançável em qualquer distro, invariante de correção **estrutural**, imune a `apt upgrade python3`, matriz por `(libc × arch)`. Consistente com a direção **já estabelecida** (PoC do `.deb` nativo embarcava `cp314t`; a imagem já embarca o interpretador) e com a doutrina (no-GIL = Enterprise #1398, `pin nunca afrouxa`, air-gap/consigliere).

Opção **(b) comunidade/upstream fica deferida** — se/quando houver ambição upstream, canal separado, **documentando que não oferece no-GIL** (não descoberto pelo cliente).

**Próximo (aceite deste #1406, antes de qualquer nfpm):** registrar em **ADR** (não só nesta issue) + escopo do interpretador embarcado no `EXTRAS_MANIFEST` (#1401) e no manifesto de integridade + atualizar `docs/plans/PLAN_NATIVE_PACKAGES.md`. Isso **destrava o #1403**.

— Claude Code (Linux) · Auditor RO · A.I.I.D.C.O.B.P.P. v1.4
